jep/1/README.adoc

```diff
@@ -0,0 +1,754 @@
+  or otherwise incorporated into the appropriate code repositories,
+  the JEP status is changed to "<<final, Final>>".  The JEP is done.
+
+. **<<maintenance, Maintenance>>** - At some later date, the JEP may need to be updated.
+  As "Author" of the JEP, Andrea makes changes as needed or hands off authorship to someone else.
+  Updates follow the same basic JEP workflow.
+
+IMPORTANT: The above is only a high-level overview of the JEP workflow.
+The full and complete description of the JEP workflow is provided below.
+Read the full description below before starting a JEP.
+
+[[start]]
+==== Start with an idea for Jenkins
+
+The JEP process begins with a new idea for Jenkins.
+A single JEP should contain a single key proposal or new idea.
+The more focused the JEP, the more successful it tends to be.
+The JEP editors reserve the right to reject potential JEPs
+if they appear too unfocused or too broad.
+If in doubt, authors should split their JEP into several well-focused ones.
+
+NOTE: Small enhancements or patches often don't need a JEP can be handle via the Jenkins
+development workflow with a JIRA item and/or pull request to the appropriate repository.
+
+==== Find a champion
+
+Each JEP must have a champion -- someone who writes the JEP using the style and
+format described below, shepherds the discussions in the appropriate forums,
+and attempts to build community consensus around the idea. The JEP champion
+(a.k.a. <<Author>>) should first attempt to ascertain whether the idea is JEP-able.
+Posting to the jenkinsci-dev@googlegroups.com mailing list is the best way to
+go about this.
+
+[[discussion]]
+==== Discuss the idea with the community
+
+Vetting an idea publicly before going as far as writing a JEP is meant
+to save the potential author time. Many ideas have been brought
+forward for changing Jenkins that have been rejected for various
+reasons. Asking the Jenkins community first if an idea is original
+helps prevent too much time being spent on something that is
+guaranteed to be rejected based on prior discussions (searching
+the internet does not always do the trick). It also helps to make sure
+the idea is applicable to the entire community and not just the author. Just
+because an idea sounds good to the author does not mean it will work for most
+people in most areas where Jenkins is used.
+
+Once the champion has asked the Jenkins community as to whether an idea has any
+chance of acceptance, a draft JEP should be presented to jenkinsci-dev@. This
+gives the author a chance to flesh out the draft JEP to make sure it is
+properly formatted, of high quality, and to address initial concerns about the
+proposal.
+
+[[submission]]
+==== Creating a JEP Submission
+
+Following a discussion on jenkinsci-dev@,
+the proposal should be turned into as a JEP submission and submitted
+via a GitHub pull request to this repository footnoteref:[repo].
```

Comment:
> This step likely means "Convert the document from a reviewable format (e.g. Google Doc) to Asciidoc". It is pretty trivial, but it still requires extra efforts. Assuming that an abstract company does internal reviews before coming up with a proposal, it means that there are likely 2 conversion steps for JEPs.

I interpret this to mean "CloudBees will have to do two conversion steps." Which I have no problems with whatsoever :smile: 

> I am not sure it's friendly to individual contributors. So such JEP process IMHO should not be required for 90% of small- and medium-scale enhancements in the core.

I definitely agree that small/medium-scale enhancements shouldn't necessarily need JEPs, but this process relies on contributor's using their good judgement to determine the scope and impact of what they're trying to accomplish, and for other contributors to helpfully guide them down a path (small -> PR, non-small impact -> JEP).

Reply:
> I interpret this to mean "CloudBees will have to do two conversion steps." Which I have no problems with whatsoever

Well, I  would expect any company-contributor to conduct an internal review before coming up with a JEP-scale proposal. 

> I definitely agree that small/medium-scale enhancements shouldn't necessarily need JEPs, but this process relies on contributor's using their good judgement to determine the scope and impact of what they're trying to accomplish, and for other contributors to helpfully guide them down a path (small -> PR, non-small impact -> JEP).

maybe also "medium -> mailing list". There are several great examples of discussions inside the Jenkins mailing list, sometimes backed by Google Doc proposal. And the difficult thing would be to determine where this approach ends and where JEP starts.

In mailing list an optimistic decision ETA is 1-2 weeks, in JEP it would be rather about 1-2 months from what I see in this document

Reply:
> In mailing list an optimistic decision ETA is 1-2 weeks, in JEP it would be rather about 1-2 months from what I see in this document

I think this is overly pessimistic. All of the well-received and well-discussed proposals ([for example](https://groups.google.com/d/msgid/jenkinsci-dev/7a7e2b81-8795-48bd-b1c2-d0ee31123df3%40googlegroups.com)), such as those using Google Docs you mention, very naturally fit into JEP in my opinion.

Creating the document shouldn't be hard, if for some reason it's more difficult to get consensus around a JEP than it is a topic on the mailing list, then I feel like that is an indication of some deeper organizational problem.


Anyways, my preference is to intentionally keep this a _bit_ ambiguous until we figure out (by practicing the process) what distinction for a JEP-able vs. a JEP-unnecessary change makes sense for us.

Reply:
agreed